### ISSUE TITLE
IS this training normal？

### ISSUE THREAD

**Issue by jingjunzhu** [CPJKU/score_following_game#16]
Hello author, sorry to bother you. I have trained the agent with nottingham for 12days and I found the learning rate is 0, is this normal?
![image](https://user-images.githubusercontent.com/95033870/166473913-508388e4-63e5-4634-9bf4-ce299f0e8240.png)

**Comment by fhenkel** [comment]
Training on Nottingham for 12 days sounds very long. But overall the avg reward looks very good, I'd consider the training to be done. I just wonder why it didn't stop earlier.
Originally, the learnrate scheduler should refine the learning rate twice (after no improvements) and then stop the training.
There might be a problem with the stopping condition in the code.

**Comment by jingjunzhu** [comment]
I have tried to evaluate the agent trained this time, and the result is much different from the pre-trained agent so I think the training not be done， but I don't know why it seems that the Learnrate is already 0.

**Comment by fhenkel** [comment]
Could you maybe share the reward plot logged to tensorboard?

**Comment by jingjunzhu** [comment]
here is the plot
![12333](https://user-images.githubusercontent.com/95033870/167239930-b6a2a166-1319-47fd-97eb-b12eac1f5181.jpg)

**Comment by fhenkel** [comment]
The learning curves look good. However, the training should have stopped once the scheduler sets the learning rate to 0.
Did you evaluate the best model or the one at the end of training/the latest iteration?

**Comment by jingjunzhu** [comment]
there is a best_model in the folder with param root, and i use "python evaluate_agent.py --trials 10 --params <PARAM-ROOT>/<run_id>/best_model.pt --data_set <PATH-TO-DATA-ROOT/nottingham/nottingham_test" to evaluate that best_model.pt, I shut down the terminal because tarining didn't stop, the result is worse than your pre-trained agent so i'm confused whether the training is successful

**Comment by fhenkel** [comment]
Are there more models in your <run_id> folder? Maybe one from an earlier update iteration, shortly before the learning rate is reduced to zero?

Other than that could you also post the evaluation results of the best one here, just to see how different they are to the trained agent.

**Comment by jingjunzhu** [comment]
each time the training start, in <rui_id> there will be two folder, but one is empty, best_model.pt will appear in another one. here is the results:
![result](https://user-images.githubusercontent.com/95033870/167470369-9601cdb0-96b2-40a2-b75b-0c7353417fde.jpg)

**Comment by jingjunzhu** [comment]
![image](https://user-images.githubusercontent.com/95033870/167470591-c6edf3f9-71a1-4401-abbd-c7179988d53b.png)
it's different from the result in paper

**Comment by fhenkel** [comment]
You are right, that should not be the case.
I'll try to train a model as well to see whether I have a similar issue.

You used the code from the tismir-2019 branch? Could you please also provide the parameters you used for the training script?

**Comment by jingjunzhu** [comment]
[2022-05.zip](https://github.com/CPJKU/score_following_game/files/8666093/2022-05.zip)
ok，here is the file

**Comment by fhenkel** [comment]
In tensorboard, there should also be a log of the exact script call. Could you please send that as well.

**Comment by jingjunzhu** [comment]
[Uploading ppo-ScoreFollowingNetNottinghamLS-nottingham_train-nottingham_ls1_20220418_135942-aibc.zip…]()
here is log

**Comment by fhenkel** [comment]
I think there is a problem with a link. Could you please check it again

**Comment by jingjunzhu** [comment]
[ppo-ScoreFollowingNetNottinghamLS-nottingham_train-nottingham_ls1_20220418_135942-aibc.zip](https://github.com/CPJKU/score_following_game/files/8687204/ppo-ScoreFollowingNetNottinghamLS-nottingham_train-nottingham_ls1_20220418_135942-aibc.zip) this maybe uploaded successfully

**Comment by fhenkel** [comment]
Thanks, that one worked. I'll come back to you once I trained a model

**Comment by fhenkel** [comment]
Alright, I think I found the issue.
First of all, the LR shown in your Tensorboard log is still >0 and the agent is still training.
The problem with the log in the console was, that I formatted the number in such way that it got rounded to 0 (not the smartest move ;)

Regarding your training in general, the default parameter for the number of PPO update epochs was not set properly.
We still had that at 4 in the code, but in our paper we just used 1 (as listed in the Appendix Table 5). This works better and will also speed up training.
However, I'd also suggest to train the agent using a GPU if possible. In that way you can achieve a good performance within a day.

**Comment by jingjunzhu** [comment]
ok! thank you, i will try to train with proper parameter, by the way, can you tell me hot wo tarin agent using GUP? i do not have idea with that.

**Comment by jingjunzhu** [comment]
i find there are some other parameters in code are different from paper, i will try to adjust them all, maybe it's useful?

**Comment by fhenkel** [comment]
To train on a GPU (if you have one available), just add the argument `--use_cuda`.
The other parameters should now already be set properly.
